File: 2017/Day5/day5.cpp
#include <iostream>
#include <string>
#include <vector>
#include <fstream>
#include <sstream>

std::vector<int> parse(std::string name){

  std::ifstream file(name);
  std::string str;
  std::vector<int> numbers;
  while(std::getline(file, str)){
    numbers.push_back(stoi(str));
  }

  return numbers;

}

int part1(std::vector<int> n){

  int steps = 0;
  int jump = 0;
  int pos = 0;

  while(pos < n.size()){
    jump = n[pos];
    n[pos]++;
    pos += jump;
    steps++;
  }

  return steps;
}

int part2(std::vector<int> n){

  int steps = 0;
  int jump = 0;
  int pos = 0;

  while(pos < n.size()){
    jump = n[pos];

    if (jump >= 3)
      n[pos]--;
    else
      n[pos]++;
    
    pos += jump;
    steps++;
  }

  return steps;
}

int main(){

  std::vector<int> numbers = parse("input.txt");
  std::cout << "Part 1: " << part1(numbers) << std::endl;
  std::cout << "Part 2: " << part2(numbers) << std::endl;

}

File: 2017/Day5/day5.rs
const INPUT: &'static str = include_str!("input.txt");


fn parse(input:&str) -> Vec<i64>{
    input.lines().map(|line| line.parse::<i64>().expect("Error"))
        .collect::<Vec<_>>()
}

fn part1(mut state: Vec<i64>) -> i64{

    let mut steps = 0;
    let mut jump:i64 = 0;
    
    while let Some(pos) = state.get_mut(jump as usize){
        jump += *pos;
        *pos += 1;
        steps += 1;
    }

    return steps;

}

fn part2(mut state: Vec<i64>) -> i64{

    let mut steps = 0;
    let mut jump:i64 = 0;
    
    while let Some(pos) = state.get_mut(jump as usize){
        jump += *pos;
        if *pos >= 3{
            *pos -= 1;
        }else{
            *pos += 1;
        }
        steps += 1;
    }

    return steps;

}

fn main(){

    let data = parse(INPUT);
    println!("Part 1: {}", part1(data.clone()));
    println!("Part 1: {}", part2(data.clone()));

}
